File: sample/rs/12_sort_visualizer.rs
use crate::time::perf_counter;
use crate::pytra::utils::gif::grayscale_palette;
use crate::pytra::utils::gif::save_gif;

// 12: Sample that outputs intermediate states of bubble sort as a GIF.

fn render(values: Vec<i64>, w: i64, h: i64) -> Vec<u8> {
    let mut frame = bytearray((w * h));
    let n = values.len() as i64;
    let bar_w = (w / n);
    let mut i: i64 = 0;
    while i < n {
        let x0 = (i * bar_w) as i64;
        let mut x1 = (((i + 1)) * bar_w) as i64;
        if x1 <= x0 {
            x1 = (x0 + 1);
        }
        let bh = (((values[i as usize] / n)) * h) as i64;
        let mut y = (h - bh);
        let mut y: i64 = y;
        while y < h {
            let mut x: i64 = x0;
            while x < x1 {
                frame[((y * w) + x) as usize] = 255;
                x += 1;
            }
            y += 1;
        }
        i += 1;
    }
    return bytes(frame);
}

fn run_12_sort_visualizer() {
    let w = 320;
    let h = 180;
    let n = 124;
    let out_path = "sample/out/12_sort_visualizer.gif";
    
    let start = perf_counter();
    let mut values: Vec<i64> = vec![];
    let mut i: i64 = 0;
    while i < n {
        values.push(((((i * 37) + 19)) % n));
        i += 1;
    }
    let mut frames: Vec<Vec<u8>> = vec![];
    let frame_stride = 16;
    
    let mut op = 0;
    let mut i: i64 = 0;
    while i < n {
        let mut swapped = false;
        let mut j: i64 = 0;
        while j < ((n - i) - 1) {
            if values[j as usize] > values[(j + 1) as usize] {
                let __tmp_1 = (values[(j + 1) as usize], values[j as usize]);
                values[j as usize] = __tmp_1.0;
                values[(j + 1) as usize] = __tmp_1.1;
                swapped = true;
            }
            if (op % frame_stride) == 0 {
                frames.push(render(values, w, h));
            }
            op += 1;
            j += 1;
        }
        if !swapped {
            py_break;
        }
        i += 1;
    }
    save_gif(out_path, w, h, frames, grayscale_palette());
    let elapsed = (perf_counter() - start);
    println!("{:?}", ("output:", out_path));
    println!("{:?}", ("frames:", frames.len() as i64));
    println!("{:?}", ("elapsed_sec:", elapsed));
}

fn main() {
    run_12_sort_visualizer();
}

File: sample/cpp/12_sort_visualizer.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/std/time.h"
#include "generated/utils/gif.h"

// 12: Sample that outputs intermediate states of bubble sort as a GIF.

bytes render(const rc<list<int64>>& values, int64 w, int64 h) {
    bytearray frame = bytearray(w * h);
    int64 n = (rc_list_ref(values)).size();
    float64 bar_w = py_to<float64>(w) / py_to<float64>(n);
    float64 __hoisted_cast_1 = float64(n);
    float64 __hoisted_cast_2 = float64(h);
    for (int64 i = 0; i < n; ++i) {
        int64 x0 = int64(py_to<float64>(i) * bar_w);
        int64 x1 = int64((py_to<float64>(i + 1)) * bar_w);
        if (x1 <= x0)
            x1 = x0 + 1;
        int64 bh = int64((py_to<float64>(py_list_at_ref(rc_list_ref(values), py_to<int64>(i))) / __hoisted_cast_1) * __hoisted_cast_2);
        int64 y = h - bh;
        auto __for_start_1 = y;
        for (int64 y = __for_start_1; y < h; ++y) {
            for (int64 x = x0; x < x1; ++x)
                frame[y * w + x] = 255;
        }
    }
    return frame;
}

void run_12_sort_visualizer() {
    int64 w = 320;
    int64 h = 180;
    int64 n = 124;
    str out_path = "sample/out/12_sort_visualizer.gif";
    
    float64 start = pytra::std::time::perf_counter();
    rc<list<int64>> values = rc_list_from_value(list<int64>{});
    rc_list_ref(values).reserve((n <= 0) ? 0 : n);
    for (int64 i = 0; i < n; ++i)
        py_list_append_mut(rc_list_ref(values), (i * 37 + 19) % n);
    rc<list<bytes>> frames = rc_list_from_value(list<bytes>{render(values, w, h)});
    int64 frame_stride = 16;
    
    int64 op = 0;
    for (int64 i = 0; i < n; ++i) {
        bool swapped = false;
        for (int64 j = 0; j < n - i - 1; ++j) {
            if (py_list_at_ref(rc_list_ref(values), py_to<int64>(j)) > py_list_at_ref(rc_list_ref(values), py_to<int64>(j + 1))) {
                ::std::swap(py_list_at_ref(rc_list_ref(values), py_to<int64>(j)), py_list_at_ref(rc_list_ref(values), py_to<int64>(j + 1)));
                swapped = true;
            }
            if (op % frame_stride == 0)
                py_list_append_mut(rc_list_ref(frames), render(values, w, h));
            op++;
        }
        if (!(swapped))
            break;
    }
    pytra::utils::gif::save_gif(out_path, w, h, rc_list_ref(frames), pytra::utils::gif::grayscale_palette(), 3, 0);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", (rc_list_ref(frames)).size());
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_12_sort_visualizer();
    return 0;
}
